libavfilter/vf_tonemap_vaapi.c

```diff
@@ -0,0 +1,454 @@
+    hdrtm_param.data.metadata_size = sizeof(VAHdrMetaDataHDR10);
+
+    return ff_vaapi_vpp_make_param_buffers(avctx,
+                                           VAProcFilterParameterBufferType,
+                                           &hdrtm_param, sizeof(hdrtm_param), 1);
+}
+
+static int tonemap_vaapi_filter_frame(AVFilterLink *inlink, AVFrame *input_frame)
+{
+    AVFilterContext *avctx     = inlink->dst;
+    AVFilterLink *outlink      = avctx->outputs[0];
+    VAAPIVPPContext *vpp_ctx   = avctx->priv;
+    HDRVAAPIContext *ctx       = avctx->priv;
+    AVFrame *output_frame      = NULL;
+    VASurfaceID input_surface, output_surface;
+
+    VAProcPipelineParameterBuffer params;
+    int err;
+
+    av_log(avctx, AV_LOG_DEBUG, "Filter input: %s, %ux%u (%"PRId64").\n",
+           av_get_pix_fmt_name(input_frame->format),
+           input_frame->width, input_frame->height, input_frame->pts);
+
+    if (vpp_ctx->va_context == VA_INVALID_ID){
+        av_frame_free(&input_frame);
+        return AVERROR(EINVAL);
+    }
+
+    err = tonemap_vaapi_save_metadata(avctx, input_frame);
+    if (err < 0)
+        goto fail;
+
+    err = tonemap_vaapi_set_filter_params(avctx, input_frame);
+    if (err < 0)
+        goto fail;
+
+    input_surface = (VASurfaceID)(uintptr_t)input_frame->data[3];
+    av_log(avctx, AV_LOG_DEBUG, "Using surface %#x for tonemap vpp input.\n",
+           input_surface);
+
+    output_frame = ff_get_video_buffer(outlink, vpp_ctx->output_width,
+                                       vpp_ctx->output_height);
+    if (!output_frame) {
+        err = AVERROR(ENOMEM);
+        goto fail;
+    }
+
+    output_surface = (VASurfaceID)(uintptr_t)output_frame->data[3];
+    av_log(avctx, AV_LOG_DEBUG, "Using surface %#x for tonemap vpp output.\n",
+           output_surface);
+    memset(&params, 0, sizeof(params));
+
+    err = av_frame_copy_props(output_frame, input_frame);
+    if (err < 0)
+        return err;
+
+    if (ctx->colour_primaries != AVCOL_PRI_UNSPECIFIED) {
+        output_frame->color_primaries = ctx->colour_primaries;
+    } else {
+        switch (vpp_ctx->output_format) {
+        case AV_PIX_FMT_NV12:
+            output_frame->color_primaries = AVCOL_PRI_BT709;
+            break;
+        default:
+            av_assert0(0);
+        }
+    }
+    if (ctx->colour_transfer != AVCOL_TRC_UNSPECIFIED) {
+        output_frame->color_trc = ctx->colour_transfer;
+    } else {
+        switch (vpp_ctx->output_format) {
+        case AV_PIX_FMT_NV12:
+            output_frame->color_trc = AVCOL_TRC_BT709;
+            break;
+        default:
+            av_assert0(0);
+        }
+    }
+    if (ctx->colour_matrix != AVCOL_SPC_UNSPECIFIED) {
+        output_frame->colorspace = ctx->colour_matrix;
+    } else {
+        switch (vpp_ctx->output_format) {
+        case AV_PIX_FMT_NV12:
+            output_frame->colorspace = AVCOL_SPC_BT709;
+            break;
```

Comment:
i don't think you need to change the output frame's colorspace, if user does not give one. just use the input's colorspace, av_frame_copy_props() already does this.